include/CL/sycl/atomic.hpp

```diff
@@ -81,16 +81,24 @@ class atomic {
 #ifdef __HIPSYCL_DEVICE_CALLABLE__
     return atomicExch(_ptr, operand);
 #else
-    return detail::invalid_host_call_dummy_return<T>();
+    return __atomic_exchange_n(_ptr, operand, __ATOMIC_RELAXED);
```

Comment:
I think the `#ifdef` code paths should be structured slightly differently. There are currently several compilation paths when compiling with hipSYCL:

* GPU compilation pass
* CPU compilation pass of the CUDA/ROCm backends. This is where we currently have those `invalid_host_call_dummy_return()` calls. Because using CPU and GPU simultaneously is not yet supported (due to limitations in the runtime which will be fixed by the new runtime), these dummy calls should never be actually executed and are in principle something like a `assert(false)`.
* CPU compilation pass when compiling for the CPU backend.

Now, the `__HIPSYCL_DEVICE_CALLABLE__` macro is defined
* by CUDA/ROCm backend if currently in the GPU compilation pass
* by the CPU backend because hipCPU defines many of those HIP backends.
In other words, it's defined when HIP intrinsics like atomicExch are available.

This means that the code here will never execute in practice because it will be compiled for the host compilation pass for the CUDA/ROCm backend.

Instead, we want to have those atomic functions run whenever we compile for CPU. Solution: Check with the `SYCL_DEVICE_ONLY` macro if we are in a GPU compilation pass:
```
#ifdef SYCL_DEVICE_ONLY
  return atomicExch(...);
#else
  return __atomic_exchange_n(...)
#endif
```

Reply:
How about now?

Reply:
Fine now, we could even drop the surrounding `#ifdef __HIPSYCL_DEVICE_CALLABLE__` to make the code simpler. There are enough other places that would prevent running CPU code in kernels when compiling for GPU. In the future I want to make the CPU backend always available (even when compiling for GPU), so then we would have to remove this anyway.